File: server/api/v1/endpoints/conversations.py
from typing import Any, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

router = APIRouter(tags=["conversations"])


def get_conversation_service():
    from server.services.conversation_service import conversation_service

    return conversation_service


class ConversationCreateRequest(BaseModel):
    participantId: str
    participantName: Optional[str] = None
    source: str = "generic"
    metadata: Optional[dict[str, Any]] = None


class MessageCreateRequest(BaseModel):
    role: str = "user"
    text: str
    metadata: Optional[dict[str, Any]] = None


class ReplyGenerateRequest(BaseModel):
    personaPrompt: str
    model: str = "gemini-2.5-flash"
    temperature: float = 0.72


class ApproveMessageRequest(BaseModel):
    messageId: str


@router.get("")
@router.get("/")
async def list_conversations():
    conversations = get_conversation_service().list_conversations()
    return {"conversations": conversations, "total": len(conversations)}


@router.post("")
@router.post("/")
async def create_conversation(request: ConversationCreateRequest):
    return get_conversation_service().create_conversation(
        participant_id=request.participantId,
        participant_name=request.participantName or request.participantId,
        source=request.source,
        metadata=request.metadata,
    )


@router.get("/{conversation_id}")
async def get_conversation(conversation_id: str):
    conversation = get_conversation_service().get_conversation(conversation_id)
    if not conversation:
        raise HTTPException(status_code=404, detail="Conversation not found")
    return conversation


@router.post("/{conversation_id}/messages")
async def add_message(conversation_id: str, request: MessageCreateRequest):
    try:
        return get_conversation_service().add_message(
            conversation_id,
            role=request.role,
            text=request.text,
            metadata=request.metadata,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc


@router.post("/{conversation_id}/reply")
async def generate_reply(conversation_id: str, request: ReplyGenerateRequest):
    try:
        return get_conversation_service().generate_reply(
            conversation_id,
            persona_prompt=request.personaPrompt,
            provider_model=request.model,
            temperature=request.temperature,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc


@router.post("/{conversation_id}/approve")
async def approve_reply(conversation_id: str, request: ApproveMessageRequest):
    try:
        return get_conversation_service().approve_message(conversation_id, request.messageId)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="Message not found") from exc

File: src/lib/conversations.ts
import { apiUrl } from './api';

export type ConversationMessage = {
  id: string;
  role: 'user' | 'assistant' | 'system';
  text: string;
  status?: 'received' | 'draft' | 'approved' | 'sent' | string;
  metadata?: Record<string, unknown>;
  createdAt: string;
  approvedAt?: string;
  sentAt?: string;
};

export type Conversation = {
  id: string;
  source: string;
  participantId: string;
  participantName: string;
  status: string;
  metadata?: Record<string, unknown>;
  messages: ConversationMessage[];
  createdAt: string;
  updatedAt: string;
};

async function parseJson<T>(response: Response): Promise<T> {
  const data = (await response.json().catch(() => ({}))) as T & { detail?: string; error?: string };
  if (!response.ok) {
    throw new Error(data.detail || data.error || `HTTP ${response.status}`);
  }
  return data;
}

export async function listConversations(): Promise<Conversation[]> {
  const data = await parseJson<{ conversations?: Conversation[] }>(
    await fetch(apiUrl('/conversations')),
  );
  return Array.isArray(data.conversations) ? data.conversations : [];
}

export async function createConversation(payload: {
  participantId: string;
  participantName?: string;
  source?: string;
  metadata?: Record<string, unknown>;
}): Promise<Conversation> {
  return parseJson<Conversation>(
    await fetch(apiUrl('/conversations'), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    }),
  );
}

export async function getConversation(conversationId: string): Promise<Conversation> {
  return parseJson<Conversation>(
    await fetch(apiUrl(`/conversations/${encodeURIComponent(conversationId)}`)),
  );
}

export async function addConversationMessage(
  conversationId: string,
  payload: {
    role?: ConversationMessage['role'];
    text: string;
    metadata?: Record<string, unknown>;
  },
): Promise<ConversationMessage> {
  return parseJson<ConversationMessage>(
    await fetch(apiUrl(`/conversations/${encodeURIComponent(conversationId)}/messages`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    }),
  );
}

export async function generateConversationReply(
  conversationId: string,
  payload: {
    personaPrompt: string;
    model?: string;
    temperature?: number;
  },
): Promise<{ message: ConversationMessage; provider: string }> {
  return parseJson<{ message: ConversationMessage; provider: string }>(
    await fetch(apiUrl(`/conversations/${encodeURIComponent(conversationId)}/reply`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    }),
  );
}

export async function approveConversationReply(
  conversationId: string,
  messageId: string,
): Promise<ConversationMessage> {
  return parseJson<ConversationMessage>(
    await fetch(apiUrl(`/conversations/${encodeURIComponent(conversationId)}/approve`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ messageId }),
    }),
  );
}
